### PR TITLE
feat(common): LOCAL-3190 Multi-lang dropdown implementation

### PR DESCRIPTION
**Pull request by bc-AntonShevchuk**
Jira: [LOCAL-3191](https://bigcommercecloud.atlassian.net/browse/LOCAL-3191)

## What/Why?
Implements a locale switcher dropdown in the storefront header, allowing users to switch between available store languages. All changes are gated behind the **LOCAL-3191.B2B_multi_language** feature flag - when disabled, the component renders nothing and no additional network requests are made.
                                                                                                                                                                                     
**Changes:**                                                                                           
New component — B3LocaleSwitcher     
  - Renders a dropdown (reusing B3DropDown) in the main header showing the active locale code
  - Determines the active locale by matching window.location.href against each locale's fullPath, with path-boundary verification to prevent partial prefix matches (e.g. /fr does not match /frozen)                                                                                    
  - Falls back to the default locale when no URL match is found   
  - Switches locale by navigating to nextLocale.fullPath + window.location.hash, preserving the current SPA route
  - Selected language persists across page reloads since the locale is derived from the URL                                                                                                                                                                                                                  
Redux store (global slice)                                                                                                                                                         
  - Added Locale type and availableLocales: Locale[] field to GlobalState                                                                                                            
  - Added setLocales action                                                                                                                                                                                                                                                            
Storefront config initialization
  - When the flag is enabled, fetches available locales via getLocales() (new GraphQL query against site.settings.locales) and dispatches setLocales 
  - Fetch failure is silently caught (.catch(() => [])) so a locale API error does not break the rest of storefront config initialization 
 
 
## Rollout/Rollback
Disable experiments:
1. LOCAL-3191.B2B_multi_language (Control of language switcher display and management of data retrieval for building the language switcher)
2. LOCAL-3235.Available_locale_fullPath (Managing the addition of a new fullPath field in Storefront GraphQL)

If disabling the experiments didn’t help.
1. Revert PR
2. Revert storefront [PR](https://github.com/bigcommerce/storefront/pull/4900)

## Testing
1. Unit tests
2. Manual test

https://github.com/user-attachments/assets/f421376f-9039-4e27-ba9f-3884b764b9b0





[LOCAL-3191]: https://bigcommercecloud.atlassian.net/browse/LOCAL-3191?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Adds a new header dropdown that changes `window.location.href` based on BigCommerce locale `fullPath`, plus a new config-time locales fetch; navigation/URL-matching bugs or misconfigured locales could affect routing when the flag is enabled.
> 
> **Overview**
> Adds a **feature-flagged** (`LOCAL-3191.B2B_multi_language`) locale switcher dropdown to the main storefront header, showing the active locale code and navigating to the selected locale while preserving the current `window.location.hash`.
> 
> Introduces a new BigCommerce GraphQL `getLocales` query and stores the returned locales in Redux (`global.locales`) via a new `setLocales` action; locales are fetched during storefront config initialization only when the flag is enabled (errors are silently ignored). Adds unit tests covering render gating, URL-to-locale matching edge cases, and navigation behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0493d90ee7dfe024064b79d01e04e19e00283ca5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

### CODE REVIEW

**Comment by RomanKrasinskyi** [comment]
@bc-AntonShevchuk Could you please resolve comments from Cursor and Copilot first. Thanks

File: apps/storefront/src/components/layout/B3Locales.test.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    renderWithProviders(<B3Locales />, {
+      preloadedState: withFlagEnabled,
+    });
+
+    expect(screen.getByRole('button', { name: /EN/i })).toBeInTheDocument();
+  });
+
+  it('shows the active locale derived from the current URL', () => {
+    vi.stubGlobal('location', { href: 'https://store.example.com/fr/some-page' });
+
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3Locales.test.tsx:78]: done

File: apps/storefront/src/components/layout/B3Locales.test.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    expect(screen.getByRole('button', { name: /FR/i })).toBeInTheDocument();
+
+    vi.unstubAllGlobals();
+  });
+
+  it('navigates to the selected locale fullPath on change', async () => {
+    const location = { href: 'https://store.example.com/#/orders', hash: '#/orders' };
+    vi.stubGlobal('location', location);
+
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3Locales.test.tsx:119]: done

File: apps/storefront/src/components/layout/B3Locales.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  if (!isMultiLocaleEnabled || availableLocales.length <= 1) {
+    return null;
+  }
+
+  const activeLocale =
+    [...availableLocales]
+      .sort((a, b) => b.fullPath.length - a.fullPath.length)
+      .find((l) => window.location.href.startsWith(l.fullPath)) ??
+    availableLocales.find((l) => l.isDefault) ??
+    availableLocales[0];
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3Locales.tsx]: done

File: apps/storefront/src/utils/storefrontConfig.ts [hunk 4]
```diff
@@ -346,6 +348,14 @@ const setStorefrontConfig = async (dispatch: DispatchProps) => {
   const { featureFlags } = store.getState().global;
   const useCombinedQuery = featureFlags['B2B-3817.disable_masquerading_cleanup_on_login'] ?? false;
 
+  let localeCode: string | undefined;
+  if (featureFlags['LOCAL-3191.B2B_multi_language']) {
+    const locales = await fetchStorefrontLocales().catch(() => []);
+    const defaultCode = locales.find((l) => l.isDefault)?.code ?? locales[0]?.code ?? 'en';
+    localeCode = getActiveLocale(locales)?.code ?? defaultCode;
+    store.dispatch(setLocales(locales));
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/utils/storefrontConfig.ts]: done

File: apps/storefront/src/components/layout/B3Locales.tsx [hunk 5]
```diff
@@ -0,0 +1,68 @@
+import { useRef, useState } from 'react';
+
+import { Box, Menu, MenuItem } from '@mui/material';
+import ListItemButton from '@mui/material/ListItemButton';
+import ListItemText from '@mui/material/ListItemText';
+
+import { useFeatureFlag } from '@/hooks/useFeatureFlag';
+import { useAppSelector } from '@/store';
+import { getActiveLocale } from '@/utils/locales';
+
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3Locales.tsx]: done

File: apps/storefront/src/utils/storefrontConfig.ts [hunk 6]
```diff
@@ -33,6 +34,7 @@ import { channelId } from '@/utils/basicConfig';
 import { FeatureFlagKey, featureFlags } from '@/utils/featureFlags';
 
 import { checkEveryPermissionsCode } from './b3CheckPermissions/check';
+import { getActiveLocale } from './locales';
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/utils/storefrontConfig.ts]: done

File: apps/storefront/src/utils/storefrontConfig.ts [hunk 7]
```diff
@@ -346,6 +348,14 @@ const setStorefrontConfig = async (dispatch: DispatchProps) => {
   const { featureFlags } = store.getState().global;
   const useCombinedQuery = featureFlags['B2B-3817.disable_masquerading_cleanup_on_login'] ?? false;
 
+  let localeCode: string | undefined;
+  if (featureFlags['LOCAL-3191.B2B_multi_language']) {
+    const locales = await fetchStorefrontLocales().catch(() => []);
+    const defaultCode = locales.find((l) => l.isDefault)?.code ?? locales[0]?.code ?? 'en';
+    localeCode = getActiveLocale(locales)?.code ?? defaultCode;
+    store.dispatch(setLocales(locales));
+  }
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/utils/storefrontConfig.ts:331]: no longer valid

File: apps/storefront/src/components/layout/B3LocaleSwitcher.tsx [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import ListItemText from '@mui/material/ListItemText';
+
+import { useFeatureFlag } from '@/hooks/useFeatureFlag';
+import { useAppSelector } from '@/store';
+import { Locale } from '@/store/slices/global';
+
+const getActiveLocale = (locales: Locale[]) =>
+  [...locales]
+    .sort((a, b) => b.fullPath.length - a.fullPath.length)
+    .find((l) => window.location.href.startsWith(l.fullPath));
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3LocaleSwitcher.tsx]: done

File: apps/storefront/src/components/layout/B3LocaleSwitcher.tsx [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+          flex: 'none',
+        },
+      }}
+    >
+      <B3DropDown
+        title={activeLocale.code.toUpperCase()}
+        list={list}
+        value={activeLocale.code}
+        handleItemClick={handleLocaleChange}
+      />
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3LocaleSwitcher.tsx]: This is a known trade-off. B3DropDown was chosen intentionally to reuse the same visual appearance (arrow indicator, font weight, colors) as the logged-in user dropdown. The disableLogoutButton behavior inside B3DropDown is a pre-existing concern unrelated to this PR.
                                                                                                                                                                                     
The correct fix - removing disableLogoutButton from B3DropDown and moving it to B3AccountInfo where it semantically belongs — is out of scope for LOCAL-3190. It should be addressed as a separate cleanup task to avoid mixing two unrelated concerns in one PR.

File: apps/storefront/src/components/layout/B3LocaleSwitcher.tsx [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+import { Locale } from '@/store/slices/global';
+
+import B3DropDown from '../B3DropDown';
+
+const getActiveLocale = (locales: Locale[]) =>
+  [...locales]
+    .sort((a, b) => b.fullPath.length - a.fullPath.length)
+    .find((l) => {
+      const { href } = window.location;
+      if (!href.startsWith(l.fullPath)) return false;
```
Comment (RomanKrasinskyi) [inline apps/storefront/src/components/layout/B3LocaleSwitcher.tsx]: 🍹 
```suggestion
      if (!href.startsWith(l.fullPath)) {
        return false;
      }
```

Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3LocaleSwitcher.tsx]: done

File: apps/storefront/src/components/layout/B3AccountInfo.tsx [hunk 11]
```diff
@@ -63,7 +63,6 @@ export default function B3AccountInfo({ closeSidebar }: B3AccountInfoProps) {
   return (
     <Box
       sx={{
-        minWidth: '150px',
```
Comment (bc-AntonShevchuk) [inline apps/storefront/src/components/layout/B3AccountInfo.tsx:66]: This is needed to remove the large margin in the place indicated in the screenshot.

Image

**Review by RomanKrasinskyi — APPROVED** [review]

File: apps/storefront/src/components/layout/B3LocaleSwitcher.tsx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+const getActiveLocale = (locales: Locales) =>
+  [...locales]
+    .sort((a, b) => b.fullPath.length - a.fullPath.length)
+    .find((l) => {
+      const { href } = window.location;
+      if (!href.startsWith(l.fullPath)) {
+        return false;
+      }
+      const next = href[l.fullPath.length];
```
Comment (funivan) [inline apps/storefront/src/components/layout/B3LocaleSwitcher.tsx:15]: why we need these statements? I think we need to have just `return href.startsWith(l.fullPath)`

File: apps/storefront/src/components/layout/B3LocaleSwitcher.tsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+  if (!isMultiLocaleEnabled || locales.length <= 1) {
+    return null;
+  }
+
+  const activeLocale = getActiveLocale(locales) ?? locales.find((l) => l.isDefault) ?? locales[0];
+
+  const list = locales.map(({ code }) => ({ key: code, name: code.toUpperCase() }));
+
+  const handleLocaleChange = (code: string | number) => {
```
Comment (funivan) [inline apps/storefront/src/components/layout/B3LocaleSwitcher.tsx:31]: in what cases we'll receive number?

File: apps/storefront/src/shared/service/bc/graphql/locales.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        isDefault
+        fullPath
+      }
+    }
+  }
+}`;
+
+export const getLocales = (): Promise<Locale[]> => {
+  const request =
+    platform === 'bigcommerce'
```
Comment (funivan) [inline apps/storefront/src/shared/service/bc/graphql/locales.ts]: could you elaborate on this, how does this work?

File: apps/storefront/src/utils/storefrontConfig.ts [hunk 15]
```diff
@@ -346,6 +348,14 @@ const setStorefrontConfig = async (dispatch: DispatchProps) => {
   const { featureFlags } = store.getState().global;
   const useCombinedQuery = featureFlags['B2B-3817.disable_masquerading_cleanup_on_login'] ?? false;
 
+  let localeCode: string | undefined;
+  if (featureFlags['LOCAL-3191.B2B_multi_language']) {
+    const locales = await fetchStorefrontLocales();
+    const defaultCode = locales.find((l) => l.isDefault)?.code ?? locales[0]?.code ?? 'en';
+    localeCode = getActiveLocale(locales)?.code ?? defaultCode;
+    store.dispatch(setLocales(locales));
+  }
```
[the remark this answers is not in the document]
Comment (funivan) [reply apps/storefront/src/utils/storefrontConfig.ts:331]: @bc-AntonShevchuk is this a legit comment?

File: apps/storefront/src/components/layout/B3LocaleSwitcher.tsx (hunk 12 above)
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3LocaleSwitcher.tsx:15]: Here I decided to be on the safe side in case, for example, href = https://sote.com/english/products, and the list of locales includes https://sote.com/en. In this case, href.startsWith(l.fullPath) would show a match.
I’m not sure this scenario is actually possible, but I feel more comfortable knowing it’s handled.

File: apps/storefront/src/components/layout/B3LocaleSwitcher.tsx (hunk 13 above)
Comment (bc-AntonShevchuk) [reply apps/storefront/src/components/layout/B3LocaleSwitcher.tsx:31]: In our case, we will never get a value of type number.

This is an interface requirement. Take a look at the [B3DropDown implementation](https://github.com/bigcommerce/b2b-buyer-portal/blob/d6ae44259cd02f3563b337a02ed9ad612d00470a/apps/storefront/src/components/B3DropDown.tsx#L129)

File: apps/storefront/src/shared/service/bc/graphql/locales.ts (hunk 14 above)
Comment (bc-AntonShevchuk) [reply apps/storefront/src/shared/service/bc/graphql/locales.ts]: I may be mistaken, please correct me if that’s the case, since I implemented this request by analogy with [getActiveBcCurrency](https://github.com/bigcommerce/b2b-buyer-portal/blob/d6ae44259cd02f3563b337a02ed9ad612d00470a/apps/storefront/src/shared/service/bc/graphql/currency.ts#L18).

I see two values for platform: default is `custom` and `bigcommerce`


[`platform = bigcommerce`](https://github.com/bigcommerce/b2b-buyer-portal/blob/825817ba83375a1fdb9f49579cc4cb64e5cd26fd/apps/storefront/src/shared/service/request/b3Fetch.ts#L144) - request used in a Stencil store to call the BigCommerce GraphQL Storefront API
[`platform = custom`](https://github.com/bigcommerce/b2b-buyer-portal/blob/825817ba83375a1fdb9f49579cc4cb64e5cd26fd/apps/storefront/src/shared/service/request/b3Fetch.ts#L154) - request used in a headless context to communicate with the BigCommerce GraphQL Storefront API via a proxy

**Comment by bc-AntonShevchuk** [comment]
cursor review

**Review by bc-akshatkaushik — APPROVED** [review]

File: apps/storefront/src/utils/storefrontConfig.ts [hunk 16]
```diff
@@ -346,6 +347,11 @@ const setStorefrontConfig = async (dispatch: DispatchProps) => {
   const { featureFlags } = store.getState().global;
   const useCombinedQuery = featureFlags['B2B-3817.disable_masquerading_cleanup_on_login'] ?? false;
 
+  if (featureFlags['LOCAL-3191.B2B_multi_language']) {
+    const locales = await getLocales().catch(() => []);
+    store.dispatch(setLocales(locales));
+  }
```
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/utils/storefrontConfig.ts:331]: done

**Comment by bc-AntonShevchuk** [comment]
cirsore review

File: apps/storefront/src/utils/storefrontConfig.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/utils/storefrontConfig.ts:331]: .catch(() => {}) added

File: apps/storefront/src/utils/storefrontConfig.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (bc-AntonShevchuk) [reply apps/storefront/src/utils/storefrontConfig.ts:331]: no longer valid

**Review by funivan — APPROVED** [review]
